src/librustc_lint/types.rs

```diff
@@ -427,42 +428,81 @@ enum FfiResult<'tcx> {
+                return false;
+            }
+
+            for field in &field_def.non_enum_variant().fields {
+                let field_ty = tcx.normalize_erasing_regions(
+                    ParamEnv::reveal_all(),
+                    field.ty(tcx, substs),
+                );
+                if is_zst(tcx, field.did, field_ty) {
+                    continue;
+                }
+
+                let attrs = tcx.get_attrs(field_def.did);
+                if attrs.iter().any(|a| a.check_name("rustc_nonnull_optimization_guaranteed")) ||
+                    ty_is_known_nonnull(tcx, field_ty) {
+                    return true;
+                }
+            }
+
+            false
+        }
+        _ => false,
+    }
+}
+
 /// Check if this enum can be safely exported based on the
 /// "nullable pointer optimization". Currently restricted
-/// to function pointers and references, but could be
-/// expanded to cover NonZero raw pointers and newtypes.
+/// to function pointers, references, core::num::NonZero*,
+/// core::ptr::NonNull, and #[repr(transparent)] newtypes.
 /// FIXME: This duplicates code in codegen.
 fn is_repr_nullable_ptr<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
-                                  def: &'tcx ty::AdtDef,
+                                  ty: Ty<'tcx>,
+                                  ty_def: &'tcx ty::AdtDef,
                                   substs: SubstsRef<'tcx>)
                                   -> bool {
-    if def.variants.len() == 2 {
-        let data_idx;
-
-        let zero = VariantIdx::new(0);
-        let one = VariantIdx::new(1);
+    if ty_def.variants.len() != 2 {
+        return false;
+    }
 
-        if def.variants[zero].fields.is_empty() {
-            data_idx = one;
-        } else if def.variants[one].fields.is_empty() {
-            data_idx = zero;
-        } else {
-            return false;
-        }
+    let get_variant_fields = |index| &ty_def.variants[VariantIdx::new(index)].fields;
+    let variant_fields = [get_variant_fields(0), get_variant_fields(1)];
+    let fields = if variant_fields[0].is_empty() {
+        &variant_fields[1]
+    } else if variant_fields[1].is_empty() {
+        &variant_fields[0]
+    } else {
+        return false;
+    };
+
+    if fields.len() != 1 {
+        return false;
+    }
 
-        if def.variants[data_idx].fields.len() == 1 {
-            match def.variants[data_idx].fields[0].ty(tcx, substs).sty {
-                ty::FnPtr(_) => {
-                    return true;
-                }
-                ty::Ref(..) => {
-                    return true;
-                }
-                _ => {}
+    let field_ty = fields[0].ty(tcx, substs);
+    let compute_size_skeleton = |t| SizeSkeleton::compute(t, tcx, ParamEnv::reveal_all());
+    match (compute_size_skeleton(ty), compute_size_skeleton(field_ty)) {
+        (Ok(ty_skeleton), Ok(field_skeleton)) => {
+            if !ty_skeleton.same_size(field_skeleton) {
```

Comment:
I think now that the cases where non-null optimization is guaranteed are explicitly enumerated, this check should always pass if `ty_is_known_nonnull` returns true. After all, if the layout is unknowable (due to generics or whatever) then we can't very well claim that the non-null optimization will actually happen.

If that seems correct to you, then I would prefer this size comparison be converted into an assertion, so that we learn (from an ICE) if for some reason it turns out to be incorrect.

Reply:
Done.